c10/core/InferenceMode.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <c10/macros/Macros.h>
+#include <c10/core/impl/LocalDispatchKeySet.h>
+
+namespace c10 {
+
+// A RAII, thread local (!) guard that enables or disables inference mode upon
+// construction, and sets it back to the original value upon destruction.
+struct TORCH_API InferenceMode {
+  InferenceMode(bool enabled=true): prev_keyset(c10::impl::tls_local_dispatch_key_set()) {
+    // Note [Expected TLS state in InferenceMode]:
+    //   InferenceMode: InplaceOrView not in included, Autograd in excluded
+    //   NormalMode: InplaceOrView in included, Autograd not in excluded
+    //
+    // 1. When InferenceMode is enabled, Autograd dispatch keys are excluded
+    //    but not InplaceOrView key.
```

Comment:
nit: This mostly repeats the comment above but in a more confusing way.
Should we remove the comment above and add the exact state here?